File: Projects/RFID Card Reader/gui/crypto1.py
"""
Crypto1 stream cipher and MIFARE Classic PRNG.

Pure Python implementation matching the AVR firmware version.
Based on the public crypto1 analysis (Garcia et al., 2008).
"""


def odd_parity8(x: int) -> int:
    """Return 1 if x has an odd number of set bits, 0 otherwise."""
    x ^= x >> 4
    x ^= x >> 2
    x ^= x >> 1
    return x & 1


def parity32(x: int) -> int:
    x ^= x >> 16
    x ^= x >> 8
    return odd_parity8(x & 0xFF)


# Filter function lookup tables
_FILTER_LUT0 = 0xF22C0044
_FILTER_LUT1 = 0x6C81

# LFSR feedback polynomial taps (split into odd/even halves)
LF_POLY_ODD = 0x29CE5C
LF_POLY_EVEN = 0x870804


def _filter_bit(odd: int) -> int:
    f = (_FILTER_LUT0 >> (odd & 0x0F)) & 1
    f |= ((_FILTER_LUT0 >> ((odd >> 4) & 0x0F)) & 1) << 1
    f |= ((_FILTER_LUT0 >> ((odd >> 8) & 0x0F)) & 1) << 2
    f |= ((_FILTER_LUT0 >> ((odd >> 12) & 0x0F)) & 1) << 3
    f |= ((_FILTER_LUT0 >> ((odd >> 16) & 0x0F)) & 1) << 4
    return (_FILTER_LUT1 >> f) & 1


class Crypto1:
    def __init__(self, key: bytes):
        """Initialize with 6-byte key."""
        k = int.from_bytes(key, "big")
        self.odd = 0
        self.even = 0
        for i in range(47, 0, -2):
            self.odd = (self.odd << 1) | ((k >> i) & 1)
        for i in range(46, -1, -2):
            self.even = (self.even << 1) | ((k >> i) & 1)

    def copy(self) -> "Crypto1":
        c = Crypto1.__new__(Crypto1)
        c.odd = self.odd
        c.even = self.even
        return c

    def crypto1_bit(self, inp: int, is_encrypted: int) -> int:
        feedin = self.odd & LF_POLY_ODD
        ret = _filter_bit(self.odd)

        feedin ^= self.even & LF_POLY_EVEN
        feedin = parity32(feedin)

        if is_encrypted:
            feedin ^= inp & 1
        else:
            feedin ^= (inp & 1) ^ ret

        self.even = ((self.even << 1) | ((self.odd >> 23) & 1)) & 0xFFFFFF
        self.odd = ((self.odd << 1) | feedin) & 0xFFFFFF

        return ret

    def crypto1_byte(self, inp: int, is_encrypted: int) -> int:
        ret = 0
        for i in range(8):
            ret |= self.crypto1_bit((inp >> i) & 1, is_encrypted) << i
        return ret

    def crypto1_word(self, inp: int, is_encrypted: int) -> int:
        ret = 0
        for i in range(32):
            ret |= self.crypto1_bit((inp >> i) & 1, is_encrypted) << i
        return ret

    def parity_check_ok(self, n: int) -> bool:
        """Check if auto-parity will be correct for next n bytes."""
        s = self.copy()
        for _ in range(n):
            ks_byte = 0
            for b in range(8):
                ks_byte |= s.crypto1_bit(0, 0) << b
            ks_par = s.crypto1_bit(0, 0)
            if odd_parity8(ks_byte) != ks_par:
                return False
        return True


def prng_successor(x: int, n: int) -> int:
    """MIFARE Classic PRNG: 32-bit LFSR successor."""
    for _ in range(n):
        x = ((x >> 1) | (
            (((x >> 16) ^ (x >> 18) ^ (x >> 19) ^ (x >> 21)) & 1) << 31
        )) & 0xFFFFFFFF
    return x

File: Projects/RFID Card Reader/src/crypto1.c
#include "crypto1.h"

// Parity lookup for 4 bits
static const uint8_t ODD_PARITY[16] = {
    0,1,1,0, 1,0,0,1, 1,0,0,1, 0,1,1,0
};

static uint8_t odd_parity8(uint8_t x) {
    return ODD_PARITY[x >> 4] ^ ODD_PARITY[x & 0x0F];
}

static uint8_t parity32(uint32_t x) {
    x ^= x >> 16;
    x ^= x >> 8;
    return odd_parity8(x & 0xFF);
}

// The nonlinear filter function
// Takes 5 bits from specific LFSR tap positions
// Uses a 20-input boolean function (factored into 4x5-bit lookups)
static const uint32_t FILTER_LUT0 = 0xF22C0044UL;
static const uint32_t FILTER_LUT1 = 0x6C81UL;

static uint8_t filter_bit(uint32_t odd) {
    uint32_t f;
    f  = FILTER_LUT0 >> (odd       & 0x0F) & 1;
    f |= (FILTER_LUT0 >> (odd >> 4  & 0x0F) & 1) << 1;
    f |= (FILTER_LUT0 >> (odd >> 8  & 0x0F) & 1) << 2;
    f |= (FILTER_LUT0 >> (odd >> 12 & 0x0F) & 1) << 3;
    f |= (FILTER_LUT0 >> (odd >> 16 & 0x0F) & 1) << 4;
    return (FILTER_LUT1 >> f) & 1;
}

// LFSR feedback polynomial taps (split into odd/even halves)
// Polynomial: x^48 + ... (see MIFARE Classic crypto analysis papers)
#define LF_POLY_ODD  0x29CE5C
#define LF_POLY_EVEN 0x870804

void crypto1_init(crypto1_state *s, uint8_t *key) {
    // Pack 6-byte key into odd/even split LFSR
    uint64_t k = 0;
    for (uint8_t i = 0; i < 6; i++) {
        k = (k << 8) | key[i];
    }

    s->odd = s->even = 0;
    for (int8_t i = 47; i > 0; i -= 2)
        s->odd = (s->odd << 1) | ((k >> i) & 1);
    for (int8_t i = 46; i >= 0; i -= 2)
        s->even = (s->even << 1) | ((k >> i) & 1);
}

uint8_t crypto1_bit(crypto1_state *s, uint8_t in, uint8_t is_encrypted) {
    uint32_t feedin = s->odd & LF_POLY_ODD;
    uint8_t ret = filter_bit(s->odd);

    feedin ^= s->even & LF_POLY_EVEN;
    feedin = parity32(feedin);

    if (is_encrypted)
        feedin ^= (in & 1);
    else
        feedin ^= (in & 1) ^ ret;

    s->even = (s->even << 1) | ((s->odd >> 23) & 1);
    s->odd = (s->odd << 1) | feedin;

    return ret;
}

uint8_t crypto1_byte(crypto1_state *s, uint8_t in, uint8_t is_encrypted) {
    uint8_t ret = 0;
    for (uint8_t i = 0; i < 8; i++) {
        ret |= crypto1_bit(s, (in >> i) & 1, is_encrypted) << i;
    }
    return ret;
}

uint32_t crypto1_word(crypto1_state *s, uint32_t in, uint8_t is_encrypted) {
    uint32_t ret = 0;
    for (uint8_t i = 0; i < 32; i++) {
        ret |= (uint32_t)crypto1_bit(s, (in >> i) & 1, is_encrypted) << i;
    }
    return ret;
}

uint32_t prng_successor(uint32_t x, uint32_t n) {
    // MIFARE Classic PRNG: 32-bit LFSR
    // Feedback: bit31 = bit0 ^ bit2 ^ bit3 ^ bit5
    // (taps at positions 16, 18, 19, 21 from MSB)
    while (n--) {
        x = (x >> 1) | (((x >> 16) ^ (x >> 18) ^ (x >> 19) ^ (x >> 21)) << 31);
    }
    return x;
}

uint8_t parity_check_ok(crypto1_state *s_copy, uint8_t n) {
    // Check if auto-parity will match for the next n encrypted bytes
    // For each byte: auto-parity is correct when odd_parity(ks_data) == ks_parity
    for (uint8_t i = 0; i < n; i++) {
        uint8_t ks_byte = 0;
        for (uint8_t b = 0; b < 8; b++) {
            ks_byte |= crypto1_bit(s_copy, 0, 0) << b;
        }
        uint8_t ks_par = crypto1_bit(s_copy, 0, 0);  // 9th bit = parity keystream
        if (odd_parity8(ks_byte) != ks_par)
            return 0;
    }
    return 1;
}
